docs_src/design/adr/0005-Service-Self-Config.md

```diff
@@ -0,0 +1,102 @@
+
+If the service finds the top-level (root) namespace is already populated with configuration information it will then read that configuration information from the configuration provider under namespace for that service (and ignore what is in the local configuration file).
+
+If the service finds the top-level (root) namespace is not populated with configuration information, it will read its local configuration file and populate the configuration provider (under the namespace for the service) with configuration read from the local configuration file.
+
+Note:  a configuration provider can be specified with a command line argument (the -cp / --configProvider) or environment variable (the edgex_configuration_provider environmental variable which overrides the command line argument).
+> NOTE:  the environmental variables are typically uppercase but there have been inconsistencies in environmental variable casing and changing it would involve non-backward compatible change.  This should be considered and made consistent in a future major release.
+
+**Using the local configuration file**
+
+When a configuration provider _isn't_ specified, the service just uses the configuration in its local configuration file.  That is the service uses the configuration in the file associated with the profile, config filename and config file directory command line options or environmental variables.  In this case, the service does not contact the configuration service (Consul) for any configuration information.
+  
+NOTE:  As the services now self seed and deployment specific changes can be made via environment overrides, it will no longer be necessary to have a Docker profile configuration file in each of the service directories (example:  https://github.com/edgexfoundry/edgex-go/blob/master/cmd/core-data/res/docker/configuration.toml).  See Consequences below.  It will still be possible for users to use the profile mechanism to specify a Docker configuration, but it will no longer be required and not the recommended approach to providing Docker container specific configuration.
+
+### Overrides
+Environmental variables override configuration or command line option values in the following ways
+- Environmental variables override local configuration; that is when configuration for a service is obtained from the local config file
+- Environmental variables override configuration values as they are pushed (self seeded) into the configuration service (Consul).  This override only occurs once (as the values are pushed / seeded into Consul from the service).  Once pushed, configuration values are used from Consul and any environmental value is ignored unless the -o/--overwrite flag is on.  The name of the environmental variable must match the path names in Consul.
```

Comment:
`any environmental value is ignored` => something like `any environmental value is ignored since already applied when pushed and now configuration service (Consul) is now the system of record for configuration.`

Reply:
I think it would be more consistent if environment variables override configuration in all situations

Reply:
@iain-anderson , once the config has been pushed into Consul with overrides applied and then you change a value (that has an override) in Consul, don't you want the changed value to be retained and not overridden the next time the service is restarted?

Reply:
No, I think if you have an override set it should always apply. If the overrides are really envisaged as for use only when migrating the configuration from file->consul, I think that would add weight to the idea that the migration should be done via another tool, eg edgex-cli. Then the tool would have the override functionality and the services would not.

Reply:
@iain-anderson , I disagree. Once Consul is populated it becomes the `system of record for configuration`.  This decision was previously made when @michaelestrin first implemented the environment variable overrides.

Reply:
I understand the reasoning behind the approach (i.e. the config provider becomes the source of truth), but I also agree with Iain, that the overrides should always apply, it just seems clunky/less obvious that they always override local configuration, but when using a config provider, only override when a services seeds it's configuration.

Reply:
Env var are the record of truth for those config items, consul is the record of truth for all others.